Fix Ticket - 57902 - REST API index does not respect fields

nirav7707 opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/57902

## Description 
This PR introduces validation for the `_links` field in the` _fields` query parameter. If the `_links` field is present, it will be appended to the response object. Otherwise, it will not be included. 

#### Test Case

###### Case 1
- URL - `/wp-json?_fields[]=name&_fields[]=_links`
- Name & links both available 

###### Case 2
- URL - `/wp-json?_fields[]=name`
- Name available 

###### Case 3
- URL - `/wp-json?_fields=name`
- Name available 

###### Case 4
- URL - `/wp-json?_fields=name,_links`
- Name & links both available 

##### Case 5
- URL - `/wp-json`
-  Entire response json object including `_links`

nirav7707 committed:
Add validation on requested fields of index end point

nirav7707 committed:
Resolve phpcs warning

nirav7707 committed:
Add default value to the link validator variable

nirav7707 committed:
Refactor code to avoid extra execution

dlh01 commented on src/wp-includes/rest-api/class-wp-rest-server.php:
```diff
@@ -565,10 +565,22 @@ public function serve_request( $path = null ) {
+		$is_link_requested = false;
 
-		if ( ! empty( $links ) ) {
+		if ( ! empty( $_GET['_fields'] ) ) {
```
`$_GET` is not going to be reliable for checking `_fields` because REST API requests can be dispatched within PHP, not just with the browser.

Also, I think relying on `$_GET` here would mean that all responses from the server would use the same set of fields, ignoring the `_fields` passed to individual requests, which I don't think is the desired outcome.

I think we'll want to use the fields that are in the request object, not globals.

nirav7707 replied:
Thank you, @dlh01 , for your feedback. I have made the necessary updates by replacing the use of `$_GET` with the `$request` object. Additionally, I have addressed the issue where the old logic failed to include the _links when the `?_fields` query variable was not passed. The updated code resolves both of these issues. Please review it on your end and let me know if any further changes are required.

dlh01 replied:
Thanks for the updates! I'll try to review the PR again this weekend.

dlh01 reviewed (commented):
Hi! Thanks for submitting the PR. I've left some notes about where I think this approach can be improved.

nirav7707 committed:
Refactor code by replacing global get to request object.

to validate the _links is requested or not in _fields query parameter, request
object is pass to function response_to_data as function parameter and assign
null as default value to the third parameter for the compatibility

nirav7707 committed:
Update function's documentation

oglekler commented on src/wp-includes/rest-api/class-wp-rest-server.php:
```diff
@@ -554,22 +554,37 @@ public function serve_request( $path = null ) {
-	public function response_to_data( $response, $embed ) {
+	public function response_to_data( $response, $embed, $request = null ) {
 		$data  = $response->get_data();
 		$links = self::get_compact_response_links( $response );
```
Is it better to check that links are needed in the first place before getting them?

kadamwhite replied:
Agreed, it would be more efficient to skip getting the links at all if we know they are not requested. That would avoid wasted effort generating a links object we don't go on to use.

nirav7707 replied:
Right, it's respecting the `_fields` in function `rest_filter_response_fields` but

```
$links = self::get_compact_response_links( $response );
if ( ! empty( $links ) ) {
	$data['_links'] = $links;
}
```

 This code will add the _links anyway, it's not validating the `_fields`. so I decide to add the $request object as parameter.

spacedmonkey commented:
Committed https://core.trac.wordpress.org/changeset/56566
